ledger/query/src/traits.rs

```diff
@@ -0,0 +1,32 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use console::{network::Network, prelude::Result, program::StatePath, types::Field};
+
+#[cfg_attr(feature = "async", async_trait)]
+pub trait QueryTrait<N: Network> {
```

Comment:
This type of "provider" trait is something that might be applicable in a few key places elsewhere in SnarkVM as well